src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java

```diff
@@ -0,0 +1,149 @@
+                LOG.info("Switching timelines. Topics: {} -> {}", activeTimeline.getTopic(), nextTimeline.getTopic());
+                switchTimeline(eventType, nextTimeline, () -> {
+                    final Timeline.StoragePosition storagePosition =
+                            StoragePositionFactory.createStoragePosition(activeTimeline, currentTopicRepo);
+                    activeTimeline.setLatestPosition(storagePosition);
+                    nextTimeline.setSwitchedAt(new Date());
+                });
+            }
+        } catch (final NakadiException ne) {
+            throw new TimelineException(ne.getMessage(), ne);
+        } catch (final InterruptedException ie) {
+            Thread.currentThread().interrupt();
+            throw new TimelineException(
+                    String.format("Timeline update was interrupted for event type %s and storage id `%s`",
+                            timelineRequest.getEventType(), timelineRequest.getStorageId()));
+        }
+    }
+
+    public Timeline getTimeline(final EventType eventType) throws TimelineException {
+        try {
+            final String eventTypeName = eventType.getName();
+            final Optional<Timeline> activeTimeline = eventTypeCache.getActiveTimeline(eventTypeName);
+            if (activeTimeline.isPresent()) {
+                return activeTimeline.get();
+            }
+
+            final Storage storage = storageDbRepository.getStorage(DEFAULT_STORAGE)
+                    .orElseThrow(() -> new NotFoundException("Fake timeline creation failed for event type " +
+                            eventType.getName() + ".No default storage defined"));
+            return Timeline.createFakeTimeline(eventType, storage);
+        } catch (final NakadiException e) {
+            LOG.error("Failed to get timeline for event type {}", eventType.getName(), e);
+            throw new TimelineException("Failed to get timeline", e);
+        }
+    }
+
+    public TopicRepository getTopicRepository(final EventType eventType)
+            throws TopicRepositoryException, TimelineException {
+        final Timeline timeline = getTimeline(eventType);
+        return topicRepositoryHolder.getTopicRepository(timeline.getStorage());
+    }
+
+    private void switchTimeline(final EventType eventType,
+                                final Timeline nextTimeline,
+                                final Runnable switcher) throws InterruptedException {
+        try {
+            timelineDbRepository.createTimeline(nextTimeline);
+            timelineSync.startTimelineUpdate(eventType.getName(), nakadiSettings.getTimelineWaitTimeoutMs());
+            switcher.run();
+            timelineDbRepository.updateTimelime(nextTimeline);
+        } catch (final Throwable th) {
+            LOG.error("Failed to switch timeline for event type {}", eventType.getName(), th);
+            timelineDbRepository.deleteTimeline(nextTimeline.getId());
```

Comment:
The actual switching happens when we call updateTimelime, that's why it should work. Another thing is that I do not update active timeline and it looks like a bug :)